// ng-scrolltop/esm2015/lib/ng-scrolltop.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, NgZone } from '@angular/core';
import { NgScrolltopService } from './ng-scrolltop.service';
/** @type {?} */
const DURATION = 1000;
export class NgScrolltopDirective {
    /**
     * @param {?} service
     * @param {?} ngZone
     */
    constructor(service, ngZone) {
        this.service = service;
        this.ngZone = ngZone;
        this.startTime = null;
        this.startTop = null;
        this.progress = 0;
        this.easing = (/**
         * @param {?} x
         * @return {?}
         */
        (x) => {
            'use strict';
            if (x < 0.5) {
                return Math.pow(x * 2, 2) / 2;
            }
            return 1 - Math.pow((1 - x) * 2, 2) / 2;
        });
        this.scrollABit = this.scrollABit.bind(this);
    }
    /**
     * @return {?}
     */
    onclick() {
        this.startTop = this.service.currentPositionY;
        this.startTime = null;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => { window.requestAnimationFrame(this.scrollABit); }));
        this.service.log('NgScrolltopDirective click');
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    scrollABit(timestamp) {
        if (!this.startTime) {
            this.startTime = timestamp;
        }
        this.progress = timestamp - this.startTime;
        /** @type {?} */
        const percent = (this.progress >= DURATION ? 1 : this.easing(this.progress / DURATION));
        /** @type {?} */
        const newY = this.startTop - Math.ceil(this.startTop * percent);
        this.service.scrollABit(newY);
        if (percent < 1) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => { window.requestAnimationFrame(this.scrollABit); }));
        }
    }
}
NgScrolltopDirective.decorators = [
    { type: Directive, args: [{
                selector: '[blScrolltop]'
            },] }
];
/** @nocollapse */
NgScrolltopDirective.ctorParameters = () => [
    { type: NgScrolltopService },
    { type: NgZone }
];
NgScrolltopDirective.propDecorators = {
    onclick: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    NgScrolltopDirective.prototype.startTime;
    /** @type {?} */
    NgScrolltopDirective.prototype.startTop;
    /** @type {?} */
    NgScrolltopDirective.prototype.progress;
    /** @type {?} */
    NgScrolltopDirective.prototype.easing;
    /** @type {?} */
    NgScrolltopDirective.prototype.service;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopDirective.prototype.ngZone;
}
