// ng-scrolltop/esm2015/lib/ng-scrolltop.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, isDevMode } from '@angular/core';
import * as i0 from "@angular/core";
/** @type {?} */
const DEFAULT_SCROLLED_PX = 500;
export class NgScrolltopService {
    constructor() {
        this.isDevMode = isDevMode();
        this.isWindow = typeof window !== 'undefined';
        this._init = false;
        // scrolling window by default or element if [elementId] is given
        this.lastY = 0;
        this.ticking = false;
        this.log = (/**
         * @param {?=} s
         * @param {...?} optional
         * @return {?}
         */
        (s, ...optional) => s && this.isDevMode && console.log(s, optional));
        this.handleScroll = (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (!this.ticking) {
                window.requestAnimationFrame((/**
                 * @return {?}
                 */
                () => {
                    this.updateLastY();
                    this.ticking = false;
                }));
                this.ticking = true;
            }
        });
    }
    /**
     * @return {?}
     */
    get currentPositionY() {
        return this._elementToScroll ? this._elementToScroll.scrollTop : window.pageYOffset;
    }
    /**
     * @return {?}
     */
    get info() { return Math.ceil(this.lastY); }
    /**
     * @return {?}
     */
    get showIcon() { return this.isWindow && this.lastY > DEFAULT_SCROLLED_PX; }
    /**
     * @return {?}
     */
    setDevMode() { this.isDevMode = true; }
    /**
     * @param {?} elementId
     * @return {?}
     */
    init(elementId) {
        if (this.isWindow) {
            this.destroy();
            this._elementToScroll = null;
            if (elementId) {
                this._elementToScroll = (/** @type {?} */ (document.querySelector(`#${elementId}`)));
            }
            this._host = this._elementToScroll || window;
            this._host.addEventListener('scroll', this.handleScroll);
            this.log(`'NgScrolltopService' to use ${this._elementToScroll ? 'element' : 'window'}`, elementId || '');
        }
        this._init = true;
    }
    /**
     * @return {?}
     */
    destroy() {
        this.isWindow && this._host && this._host.removeEventListener('scroll', this.handleScroll);
    }
    /**
     * @param {?} newY
     * @return {?}
     */
    scrollABit(newY) {
        if (this._elementToScroll) {
            this._elementToScroll.scrollTop = newY;
        }
        else {
            window.scroll(0, newY);
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateLastY() {
        if (!this._init)
            return;
        /** @type {?} */
        const newY = this.currentPositionY;
        if (newY !== this.lastY) {
            this.lastY = newY;
        }
    }
}
NgScrolltopService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgScrolltopService.ctorParameters = () => [];
/** @nocollapse */ NgScrolltopService.ngInjectableDef = i0.defineInjectable({ factory: function NgScrolltopService_Factory() { return new NgScrolltopService(); }, token: NgScrolltopService, providedIn: "root" });
if (false) {
    /** @type {?} */
    NgScrolltopService.prototype.isDevMode;
    /** @type {?} */
    NgScrolltopService.prototype.isWindow;
    /** @type {?} */
    NgScrolltopService.prototype._init;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype._elementToScroll;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype._host;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype.lastY;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype.ticking;
    /** @type {?} */
    NgScrolltopService.prototype.log;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype.handleScroll;
}
