// ng-scrolltop/esm5/lib/ng-scrolltop.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, NgZone } from '@angular/core';
import { NgScrolltopService } from './ng-scrolltop.service';
/** @type {?} */
var DURATION = 1000;
var NgScrolltopDirective = /** @class */ (function () {
    function NgScrolltopDirective(service, ngZone) {
        this.service = service;
        this.ngZone = ngZone;
        this.startTime = null;
        this.startTop = null;
        this.progress = 0;
        this.easing = (/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            'use strict';
            if (x < 0.5) {
                return Math.pow(x * 2, 2) / 2;
            }
            return 1 - Math.pow((1 - x) * 2, 2) / 2;
        });
        this.scrollABit = this.scrollABit.bind(this);
    }
    /**
     * @return {?}
     */
    NgScrolltopDirective.prototype.onclick = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.startTop = this.service.currentPositionY;
        this.startTime = null;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () { window.requestAnimationFrame(_this.scrollABit); }));
        this.service.log('NgScrolltopDirective click');
    };
    /**
     * @param {?} timestamp
     * @return {?}
     */
    NgScrolltopDirective.prototype.scrollABit = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        var _this = this;
        if (!this.startTime) {
            this.startTime = timestamp;
        }
        this.progress = timestamp - this.startTime;
        /** @type {?} */
        var percent = (this.progress >= DURATION ? 1 : this.easing(this.progress / DURATION));
        /** @type {?} */
        var newY = this.startTop - Math.ceil(this.startTop * percent);
        this.service.scrollABit(newY);
        if (percent < 1) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { window.requestAnimationFrame(_this.scrollABit); }));
        }
    };
    NgScrolltopDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[blScrolltop]'
                },] }
    ];
    /** @nocollapse */
    NgScrolltopDirective.ctorParameters = function () { return [
        { type: NgScrolltopService },
        { type: NgZone }
    ]; };
    NgScrolltopDirective.propDecorators = {
        onclick: [{ type: HostListener, args: ['click',] }]
    };
    return NgScrolltopDirective;
}());
export { NgScrolltopDirective };
if (false) {
    /** @type {?} */
    NgScrolltopDirective.prototype.startTime;
    /** @type {?} */
    NgScrolltopDirective.prototype.startTop;
    /** @type {?} */
    NgScrolltopDirective.prototype.progress;
    /** @type {?} */
    NgScrolltopDirective.prototype.easing;
    /** @type {?} */
    NgScrolltopDirective.prototype.service;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopDirective.prototype.ngZone;
}
