// ng-scrolltop/esm5/lib/ng-scrolltop.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, isDevMode } from '@angular/core';
import * as i0 from "@angular/core";
/** @type {?} */
var DEFAULT_SCROLLED_PX = 500;
var NgScrolltopService = /** @class */ (function () {
    function NgScrolltopService() {
        var _this = this;
        this.isDevMode = isDevMode();
        this.isWindow = typeof window !== 'undefined';
        this._init = false;
        // scrolling window by default or element if [elementId] is given
        this.lastY = 0;
        this.ticking = false;
        this.log = (/**
         * @param {?=} s
         * @param {...?} optional
         * @return {?}
         */
        function (s) {
            var optional = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                optional[_i - 1] = arguments[_i];
            }
            return s && _this.isDevMode && console.log(s, optional);
        });
        this.handleScroll = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (!_this.ticking) {
                window.requestAnimationFrame((/**
                 * @return {?}
                 */
                function () {
                    _this.updateLastY();
                    _this.ticking = false;
                }));
                _this.ticking = true;
            }
        });
    }
    Object.defineProperty(NgScrolltopService.prototype, "currentPositionY", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementToScroll ? this._elementToScroll.scrollTop : window.pageYOffset;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgScrolltopService.prototype, "info", {
        get: /**
         * @return {?}
         */
        function () { return Math.ceil(this.lastY); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgScrolltopService.prototype, "showIcon", {
        get: /**
         * @return {?}
         */
        function () { return this.isWindow && this.lastY > DEFAULT_SCROLLED_PX; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgScrolltopService.prototype.setDevMode = /**
     * @return {?}
     */
    function () { this.isDevMode = true; };
    /**
     * @param {?} elementId
     * @return {?}
     */
    NgScrolltopService.prototype.init = /**
     * @param {?} elementId
     * @return {?}
     */
    function (elementId) {
        if (this.isWindow) {
            this.destroy();
            this._elementToScroll = null;
            if (elementId) {
                this._elementToScroll = (/** @type {?} */ (document.querySelector("#" + elementId)));
            }
            this._host = this._elementToScroll || window;
            this._host.addEventListener('scroll', this.handleScroll);
            this.log("'NgScrolltopService' to use " + (this._elementToScroll ? 'element' : 'window'), elementId || '');
        }
        this._init = true;
    };
    /**
     * @return {?}
     */
    NgScrolltopService.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this.isWindow && this._host && this._host.removeEventListener('scroll', this.handleScroll);
    };
    /**
     * @param {?} newY
     * @return {?}
     */
    NgScrolltopService.prototype.scrollABit = /**
     * @param {?} newY
     * @return {?}
     */
    function (newY) {
        if (this._elementToScroll) {
            this._elementToScroll.scrollTop = newY;
        }
        else {
            window.scroll(0, newY);
        }
    };
    /**
     * @private
     * @return {?}
     */
    NgScrolltopService.prototype.updateLastY = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this._init)
            return;
        /** @type {?} */
        var newY = this.currentPositionY;
        if (newY !== this.lastY) {
            this.lastY = newY;
        }
    };
    NgScrolltopService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    NgScrolltopService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgScrolltopService.ngInjectableDef = i0.defineInjectable({ factory: function NgScrolltopService_Factory() { return new NgScrolltopService(); }, token: NgScrolltopService, providedIn: "root" });
    return NgScrolltopService;
}());
export { NgScrolltopService };
if (false) {
    /** @type {?} */
    NgScrolltopService.prototype.isDevMode;
    /** @type {?} */
    NgScrolltopService.prototype.isWindow;
    /** @type {?} */
    NgScrolltopService.prototype._init;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype._elementToScroll;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype._host;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype.lastY;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype.ticking;
    /** @type {?} */
    NgScrolltopService.prototype.log;
    /**
     * @type {?}
     * @private
     */
    NgScrolltopService.prototype.handleScroll;
}
